sycl/source/detail/kernel_bundle_impl.hpp

```diff
@@ -351,30 +351,73 @@ class kernel_bundle_impl {
   kernel
   get_kernel(const kernel_id &KernelID,
              const std::shared_ptr<detail::kernel_bundle_impl> &Self) const {
+    using ImageImpl = std::shared_ptr<detail::device_image_impl>;
+    // Selected image.
+    ImageImpl SelectedImage = nullptr;
+    // Image where specialization constants are replaced with default values.
+    ImageImpl ImageWithReplacedSpecConsts = nullptr;
+    // Original image where specialization constants are not replaced with
+    // default values.
+    ImageImpl OriginalImage = nullptr;
+    // Used to track if any of the candidate images has specialization values
+    // set.
+    bool SpecConstsSet = false;
+    for (auto &DeviceImage : MDeviceImages) {
+      if (!DeviceImage.has_kernel(KernelID))
+        continue;
+
+      const auto DeviceImageImpl = detail::getSyclObjImpl(DeviceImage);
+      SpecConstsSet |= DeviceImageImpl->is_any_specialization_constant_set();
+
+      // Remember current image in corresponding variable depending on whether
+      // specialization constants are replaced with default value or not.
+      (DeviceImageImpl->specialization_constants_replaced_with_default()
+           ? ImageWithReplacedSpecConsts
+           : OriginalImage) = DeviceImageImpl;
+
+      if (SpecConstsSet) {
+        // If specialization constant is set in any of the candidate images
+        // then we can't use ReplacedImage, so we select NativeImage if any or
+        // we select OriginalImage and keep iterating in case there is an image
+        // with native support.
+        SelectedImage = OriginalImage ? OriginalImage : nullptr;
+        if (SelectedImage &&
+            SelectedImage->all_specialization_constant_native())
+          break;
+
+        continue;
+      } else {
+        // For now select ReplacedImage but it may be reset if any of the
+        // further device images has specialization constant value set. If after
+        // all iterations specialization constant values are not set in any of
+        // the candidate images then that will be the selected image.
+        // Also we don't want to use ReplacedImage if device image has native
+        // support.
+        if (ImageWithReplacedSpecConsts &&
+            !ImageWithReplacedSpecConsts->all_specialization_constant_native())
+          SelectedImage = ImageWithReplacedSpecConsts;
+        else
+          // In case if we don't have or don't use ReplacedImage.
+          SelectedImage = OriginalImage;
 
-    auto It = std::find_if(MDeviceImages.begin(), MDeviceImages.end(),
-                           [&KernelID](const device_image_plain &DeviceImage) {
-                             return DeviceImage.has_kernel(KernelID);
-                           });
+        continue;
+      }
+    }
```

Comment:
I wonder if there are any other early-exit conditions we could use to avoid looking for a kernel in images that we know won't have it or won't be a better option for the selection?

Reply:
Unfortunately, I didn't came up with any other early-exit conditions right now. Since we skip images that don't have our kernel, I expect that in standard situations there will be not so many images that we actually process.